tests/two_d_primitive_tests.cpp

```diff
@@ -0,0 +1,106 @@
+    {0xFFFFFF, 0, 0, 799, 599},
+    {0xFFFFFF, 799, 599, 0, 0},
+    {0xFFFFFF, 400, 0, 400, 599},
+    {0xFFFFFF, 400, 300, 401, 300},
+    {0xFFFFFF, 799, 599, 0, 0},
+    {0xFFFFFF, 999, 300, 0, 300},
+};
+
+TwoDPrimitiveTests::TwoDPrimitiveTests(TestHost& host, std::string output_dir)
+    : TestSuite(host, std::move(output_dir)) {
+  for (auto test : kTests) {
+    std::string name = MakeTestName(test);
+
+    auto test_method = [this, test]() { this->Test(test); };
+    tests_[name] = test_method;
+  }
+}
+
+void TwoDPrimitiveTests::Initialize() {
+  SetDefaultTextureFormat();
+
+  auto channel = kNextContextChannel;
+
+  pb_create_gr_ctx(channel++, NV04_SOLID_LINE, &solid_lin_ctx_);
+  pb_bind_channel(&solid_lin_ctx_);
+  pb_bind_subchannel(SUBCH_CLASS_5C, &solid_lin_ctx_);
+
+  pb_create_gr_ctx(channel++, NV04_CONTEXT_SURFACES_2D, &surface_destination_ctx_);
+  pb_bind_channel(&surface_destination_ctx_);
+  pb_bind_subchannel(SUBCH_CLASS_42, &surface_destination_ctx_);
+
+  // set up to not have shader
+  // no vertex shader (transform shader)..
+  // no fragment or pixel shader (color from triangle)..
+  host_.SetShaderProgram(nullptr);
+}
+
+void TwoDPrimitiveTests::Test(const TwoDPrimitiveTest& test) {
+  host_.PrepareDraw(0xFF440011); //beta + RRGGBB
+
+  auto p = pb_begin();
+
+  p = pb_push1_to(SUBCH_CLASS_42, p, NV04_CONTEXT_SURFACES_2D_SET_DMA_IMAGE_DST, DMA_CHANNEL_FOR_PIXEL_RENDERER_9);
+
+#define NV042_SET_PITCH 0x304
+  p = pb_push1_to(SUBCH_CLASS_42, p, NV042_SET_PITCH, (pb_back_buffer_pitch() << 16) | pb_back_buffer_pitch());
+
+#define NV042_SET_COLOR_FORMAT NV04_CONTEXT_SURFACES_2D_FORMAT
+#define NV042_SET_COLOR_FORMAT_LE_A8R8G8B8 0x0A
+  p = pb_push1_to(SUBCH_CLASS_42, p, NV042_SET_COLOR_FORMAT, NV042_SET_COLOR_FORMAT_LE_A8R8G8B8);
+
+  p = pb_push1_to(SUBCH_CLASS_5C, p, NV04_SOLID_LINE_OPERATION, NV09F_SET_OPERATION_SRCCOPY);
+  p = pb_push1_to(SUBCH_CLASS_5C, p, NV04_SOLID_LINE_COLOR_VALUE, test.two_d_object_color);
+  p = pb_push1_to(SUBCH_CLASS_5C, p, NV04_SOLID_LINE_SURFACE, surface_destination_ctx_.ChannelID);
+
+  // TODO: test to learn if 0x3 really is NV097_SET_SURFACE_FORMAT_COLOR_LE_R5G6B5
+  p = pb_push1_to(SUBCH_CLASS_5C, p, NV04_SOLID_LINE_COLOR_FORMAT, 0x3);
+  p = pb_push1_to(SUBCH_CLASS_5C, p, NV04_SOLID_LINE_START, (test.two_d_start_y << 16) | test.two_d_start_x);
+  p = pb_push1_to(SUBCH_CLASS_5C, p, NV04_SOLID_LINE_END, (test.two_d_end_y << 16) | test.two_d_end_y);
+
+  pb_end(p);
+
+  pb_print("2D Line: (%lu, %lu) - (%lu, %lu)\n", test.two_d_start_x, test.two_d_start_y, test.two_d_end_x, test.two_d_end_y);
+  pb_print("Color: %08X\n", test.two_d_object_color);
+
+  pb_draw_text_screen();
+  //  std::string name = MakeTestName(test);
+
+  host_.FinishDraw();
+  // host_.FinishDrawAndSave(output_dir_.c_str(), name.c_str());
+}
+
+std::string TwoDPrimitiveTests::MakeTestName(const TwoDPrimitiveTest& test) {
+  char buf[256] = {0};
+  snprintf(buf, 255, "2D Line - %08X - Sx%lu Sy%lu - Ex%lu Ey%lu",
```

Comment:
This might end up generating a filename that is too long and will blow up when `FinishDrawAndSave` is called. The current convention is also to not have any spaces in the filename, so I'd do underscores and be really terse. 

As an aside, once #8 is done we can probably revisit naming and reclaim some of the chars used for test suite names.

Reply:
done